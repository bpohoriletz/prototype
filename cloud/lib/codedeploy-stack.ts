import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';

import { createCdApplication } from './aws-cdk-kit/codedeploy/app';
import { createCodedeployRole } from './aws-cdk-kit/codedeploy/role';
import { createPublicDeploymentGroup } from './aws-cdk-kit/codedeploy/deployment-group';
import { createEc2InstanceProfile } from './aws-cdk-kit/roles/ec2-profile';
// import { createGithubCliRole } from './aws-cdk-kit/iam/role';
import { createTempPrivateBucket } from './aws-cdk-kit/s3/bucket';
import PetVpcBuilder from './aws-cdk-kit/ec2/vpc-builders/pet';
import VpcDirector from './aws-cdk-kit/directors/vpc';

export class PrototypeStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
  }

  public async synth(stackName: string, projectName: string) {
    const resourceNamePrefix = [stackName, projectName];
    // Step 1: Get VPC
    const parserVpc = new VpcDirector(PetVpcBuilder).constructVpc(this, 'Vpc');
    // Step 2: Create S3 bucket
    const cdBucket = createTempPrivateBucket([`codedeploy-${stackName}-${projectName}`, this.region, this.account], this);
    // Step 3: Create Roles
    // createGithubCliRole('bpohoriletz', this);
    const cdRole = createCodedeployRole(resourceNamePrefix, cdBucket.bucketArn, this);
    const instanceProfile = createEc2InstanceProfile(resourceNamePrefix, cdBucket.bucketArn, this);
    // Step 4: Create App
    const parserApp = createCdApplication(resourceNamePrefix, this);
    parserApp.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY);
    // Step 5: Create Deployment Group
    const instanceTypes = [new ec2.InstanceType('t2.micro')];
    createPublicDeploymentGroup(resourceNamePrefix, instanceTypes, parserApp, instanceProfile, cdRole, parserVpc, this);
  }
}
