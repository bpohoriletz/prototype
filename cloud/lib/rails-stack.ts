import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { Bucket } from "aws-cdk-lib/aws-s3";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as ssm from "aws-cdk-lib/aws-ssm";

import { createAppRole } from "./aws-cdk-kit/roles/app";
import { createApplication } from "./aws-cdk-kit/eb/app";
import { createEc2InstanceProfile } from "./aws-cdk-kit/roles/ec2-profile";
import { createEnvironment } from "./aws-cdk-kit/eb/env";
import { createInitAppVersions } from "./aws-cdk-kit/eb/app-version";
import { createPrivateBucket } from "./aws-cdk-kit/s3/bucket";
import * as con from "./aws-cdk-kit/naming/resources";

export class RailsStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
  }

  public async synth(stackName: string, projectName: string) {
    const environmentName = "demo";
    const resourceNamePrefix = [stackName, projectName, environmentName];
    // const resourceTags: string[] = [this.stackId, this.region, this.account];
    // Step 0 (tricky): Create regional bucket for ElasticBeanstalk
    const regionalEbBucket = Bucket.fromBucketName(this, "RegionalEbBucket", `elasticbeanstalk-${this.region}-${this.account}`);
    // Step 1: Fetch VPC
    const vpcID = ssm.StringParameter.fromStringParameterName(this, "vpcID", `/${stackName}/VpcID`).stringValue;
    const preProductionVpc = ec2.Vpc.fromLookup(this, "RailsVpc", {vpcName: con.ec2VpcName([stackName])})
    // Step H: Fetch RDS data
    // Step 2: Create ElasticBeanstalk application
    const appRole = createAppRole(resourceNamePrefix, this);
    const preProductionApp = createApplication(resourceNamePrefix, appRole, this);
    // Step 3: Create ElasticBeanstalk environment
    const instanceProfile = createEc2InstanceProfile(resourceNamePrefix, [regionalEbBucket.bucketArn], this);
    const [demoEnv, _demoSg] = await createEnvironment(preProductionApp, resourceNamePrefix, instanceProfile, preProductionVpc, this, "64bit Amazon Linux 2023 v4.4.0 running Ruby 3.4");
    // Step 4 (optional): Create S3 bucket for ElasticBeanstalk environment
    const _ebBucket = createPrivateBucket(resourceNamePrefix, this);
    // Step 5: Create application version
    const appVersion = createInitAppVersions(resourceNamePrefix, preProductionApp, regionalEbBucket.bucketArn, this);
    // Step 6(optional): Deploy application version to ElasticBeanstalk environment
    if (this.node.tryGetContext("deployInitialVersion") == "yes") {
      demoEnv.versionLabel = appVersion.ref;
    }
  }
}
