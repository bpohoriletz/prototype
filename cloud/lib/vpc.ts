import { Stack } from "aws-cdk-lib";
import { Vpc, IpAddresses, SubnetConfiguration, SubnetType } from "aws-cdk-lib/aws-ec2";
import * as con from "./naming/resources";

export function createMinimalVpc(resourceNamePrefix: string[], stack: Stack) : Vpc {
    // configure subnets
    const subnetConfigurations: SubnetConfiguration[] = [
      {
        subnetType: SubnetType.PUBLIC,
        name: con.ec2PublicSubnetName(resourceNamePrefix),
      },
      {
        subnetType: SubnetType.PRIVATE_ISOLATED,
        name: con.ec2IsolatedSubnetName(resourceNamePrefix),
      },
      {
        subnetType: SubnetType.PRIVATE_WITH_EGRESS,
        name: con.ec2PrivateSubnetName(resourceNamePrefix),
      }
    ];
    // create vpc
    const vpc = new Vpc(stack, con.ec2VpcName(resourceNamePrefix), {
      ipAddresses: IpAddresses.cidr("10.0.0.0/16"),
      natGateways: 1,
      subnetConfiguration: subnetConfigurations,
      maxAzs: 2,
    });

    return vpc;
}
