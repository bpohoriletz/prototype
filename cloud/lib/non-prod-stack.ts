import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
// import * as ssm from "aws-cdk-lib/aws-ssm";
// import * as s3 from "aws-cdk-lib/aws-s3";

// import { createNonprodDatabase } from "./aws-cdk-kit/rds/postgres"
import { createVpc } from "./aws-cdk-kit/ec2/vpc";

export class NonProdStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
  }

  public async synth(stackName: string) {
    const resourceNamePrefix = [stackName];
    // Step 1: Create VPC
    const _nonProdVpc = createVpc(resourceNamePrefix, this);
    // Step H: Import ElasticBeanstalk bucket
    // const _sharedVpcId = new ssm.StringParameter(this, `${stackName}VpcID`, {
    //   parameterName: `/${stackName}/VpcID`,
    //   stringValue: nonProdVpc.vpcId,
    //   description: "This parameter stores a configuration value for my application",
    // });
    // Step H: Import ElasticBeanstalk bucket
    // const regionalEbBucket = s3.Bucket.fromBucketName(this, "RegionalEbBucket", `elasticbeanstalk-${this.region}-${this.account}`);
    // Step H: Create RDS server
    // const [_preProductionDb, _rdsSg ] = createNonprodDatabase(resourceNamePrefix, this, preProductionVpc);
  }
}
