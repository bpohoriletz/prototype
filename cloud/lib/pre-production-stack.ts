import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { Vpc } from "aws-cdk-lib/aws-ec2";
import {
  CfnApplication,
  CfnApplicationVersion,
  CfnEnvironment,
} from "aws-cdk-lib/aws-elasticbeanstalk";
import { CfnInstanceProfile, Role } from "aws-cdk-lib/aws-iam";
import { CfnBucket } from "aws-cdk-lib/aws-s3";
import { DatabaseInstance } from "aws-cdk-lib/aws-rds";

import { createMinimalVpc } from "./vpc"
import { createRegionalEbBucket } from "./s3/bucket"
import { createAppRole } from "./roles/app"
import { createApplication } from "./eb/app"
import { createInitAppVersions } from "./eb/app-version"
import { createEc2InstanceProfile } from "./roles/ec2-profile"
import { createEnvironment } from "./eb/env"
import { createDatabase } from "./rds/postgres"

export class PreProductionStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
    const projectName = "prototype";
    const stackName: string = id.replace("Stack", "");
    const environmentName = "demo";
    // dirrerent environments can't be specified with parameters
    // const environmentName: string = new CfnParameter(this, "environmentName", { type: "String", default: "demo" }).valueAsString.toLowerCase();
    const resourceNamePrefix: string[] = [projectName, stackName, environmentName];
    // const resourceTags: string[] = [this.stackId, this.region, this.account];
    // Step 0 (tricky): Create regional bucket for ElasticBeanstalk
    const regionalEbBucket: CfnBucket = createRegionalEbBucket(this);
    // Step 1: Create VPC
    const preProductionVpc: Vpc = createMinimalVpc(resourceNamePrefix, this);
    // Step 2: Create RDS server
    const preProductionDb: DatabaseInstance = createDatabase(resourceNamePrefix, this, preProductionVpc);
    // Step 3: Create ElasticBeanstalk application
    const appRole: Role = createAppRole(resourceNamePrefix, this);
    const preProductionApp: CfnApplication = createApplication(resourceNamePrefix, appRole, this);
    // Step 3: Create ElasticBeanstalk environment
    const instanceProfile: CfnInstanceProfile = createEc2InstanceProfile(resourceNamePrefix, regionalEbBucket.attrArn, this);
    const demoEnv: CfnEnvironment = createEnvironment(preProductionApp, resourceNamePrefix, preProductionVpc, instanceProfile, preProductionDb, this);
    // Step 4 (optional): Create S3 bucket for ElasticBeanstalk environment
    // const demoAppBucket = createPrivateBucket(resourceNamePrefix, this);
    // Step 5: Create application version
    const appVersion: CfnApplicationVersion = createInitAppVersions(resourceNamePrefix, preProductionApp, regionalEbBucket, this)
    // Step 6(optional): Deploy application version to ElasticBeanstalk environment
    if (this.node.tryGetContext("deployInitialVersion") == "yes") {
      demoEnv.versionLabel = appVersion.ref
    }
  }
}
