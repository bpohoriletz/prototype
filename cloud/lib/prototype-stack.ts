import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { Bucket } from "aws-cdk-lib/aws-s3";

import { createAppRole } from "./aws-cdk-kit/roles/app"
import { createApplication } from "./aws-cdk-kit/eb/app"
import { createNonprodDatabase } from "./aws-cdk-kit/rds/postgres"
import { createEc2InstanceProfile } from "./aws-cdk-kit/roles/ec2-profile"
import { createEnvironment } from "./aws-cdk-kit/eb/env"
import { createInitAppVersions } from "./aws-cdk-kit/eb/app-version"
import { createMinimalVpc } from "./aws-cdk-kit/ec2/vpc"
import { createPrivateBucket } from "./aws-cdk-kit/s3/bucket";

export class PrototypeStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
  }

  public async synth(stackName: string, projectName: string) {
    const environmentName = "demo";
    const resourceNamePrefix = [projectName, stackName, environmentName];
    // const resourceTags: string[] = [this.stackId, this.region, this.account];
    // Step 0 (tricky): Create regional bucket for ElasticBeanstalk
    const regionalEbBucket = Bucket.fromBucketName(this, "RegionalEbBucket", `elasticbeanstalk-${this.region}-${this.account}`);
    // Step 1: Create VPC
    const preProductionVpc = createMinimalVpc(resourceNamePrefix, this);
    // Step 2: Create RDS server
    const [_preProductionDb, _rdsSg ] = createNonprodDatabase(resourceNamePrefix, this, preProductionVpc);
    // Step 3: Create ElasticBeanstalk application
    const appRole = createAppRole(resourceNamePrefix, this);
    const preProductionApp = createApplication(resourceNamePrefix, appRole, this);
    // Step 3: Create ElasticBeanstalk environment
    const instanceProfile = createEc2InstanceProfile(resourceNamePrefix, [regionalEbBucket.bucketArn], this);
    const [demoEnv, _demoSg] = await createEnvironment(preProductionApp, resourceNamePrefix, instanceProfile, preProductionVpc, this, "64bit Amazon Linux 2023 v4.4.0 running Ruby 3.4");
    // Step 4 (optional): Create S3 bucket for ElasticBeanstalk environment
    const _ebBucket = createPrivateBucket(resourceNamePrefix, this);
    // Step 5: Create application version
    const appVersion = createInitAppVersions(resourceNamePrefix, preProductionApp, regionalEbBucket.bucketArn, this)
    // Step 6(optional): Deploy application version to ElasticBeanstalk environment
    if (this.node.tryGetContext("deployInitialVersion") == "yes") {
      demoEnv.versionLabel = appVersion.ref
    }
  }
}
