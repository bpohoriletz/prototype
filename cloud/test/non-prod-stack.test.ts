import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";

import { NonProdStack } from "../lib/non-prod-stack";

describe("new NonProdStack()", () => {
  let stack: NonProdStack;

  beforeEach(() => {
    const app = new cdk.App();
    stack = new NonProdStack(app, "NonProdStackID", { env: { account: "00000000", region: "us-east-1"} })
  });

  test("matches snapshot", async () => {
    await stack.synth("stackName");

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
