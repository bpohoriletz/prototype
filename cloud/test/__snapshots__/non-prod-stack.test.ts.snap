// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new NonProdStack() matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "StackNameVpc5540D0A5": {
      "Properties": {
        "CidrBlock": "10.255.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "StackNameVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "StackNameVpcIGW709ECB36": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "StackNameVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "StackNameVpcPublicSubnet1DefaultRoute034C2FEC": {
      "DependsOn": [
        "StackNameVpcVPCGW9AA1996E",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "StackNameVpcIGW709ECB36",
        },
        "RouteTableId": {
          "Ref": "StackNameVpcPublicSubnet1RouteTableBEE95E63",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "StackNameVpcPublicSubnet1RouteTableAssociation2CE66F07": {
      "Properties": {
        "RouteTableId": {
          "Ref": "StackNameVpcPublicSubnet1RouteTableBEE95E63",
        },
        "SubnetId": {
          "Ref": "StackNameVpcPublicSubnet1SubnetAF92EB52",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "StackNameVpcPublicSubnet1RouteTableBEE95E63": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "NonProdStackID/StackNameVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "StackNameVpc5540D0A5",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "StackNameVpcPublicSubnet1SubnetAF92EB52": {
      "Properties": {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.255.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "NonProdStackID/StackNameVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "StackNameVpc5540D0A5",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "StackNameVpcPublicSubnet2DefaultRouteF1010FF7": {
      "DependsOn": [
        "StackNameVpcVPCGW9AA1996E",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "StackNameVpcIGW709ECB36",
        },
        "RouteTableId": {
          "Ref": "StackNameVpcPublicSubnet2RouteTableF266821F",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "StackNameVpcPublicSubnet2RouteTableAssociation8C7EAE6B": {
      "Properties": {
        "RouteTableId": {
          "Ref": "StackNameVpcPublicSubnet2RouteTableF266821F",
        },
        "SubnetId": {
          "Ref": "StackNameVpcPublicSubnet2Subnet700BE41F",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "StackNameVpcPublicSubnet2RouteTableF266821F": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "NonProdStackID/StackNameVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "StackNameVpc5540D0A5",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "StackNameVpcPublicSubnet2Subnet700BE41F": {
      "Properties": {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.255.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "NonProdStackID/StackNameVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "StackNameVpc5540D0A5",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "StackNameVpcVPCGW9AA1996E": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "StackNameVpcIGW709ECB36",
        },
        "VpcId": {
          "Ref": "StackNameVpc5540D0A5",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
